src/metalookup/features/addblock_based.py

```diff
@@ -1,9 +1,127 @@
-from core.extractor import ExtractionMethod, MetadataBase, ProbabilityDeterminationMethod
-
-
-@MetadataBase.with_key()
-class Advertisement(MetadataBase):
-    urls = [
+import asyncio
+import logging
+from concurrent.futures import Executor
+from logging import Logger
+from typing import Optional
+from unittest import mock
+
+import adblockparser
+from aiohttp import ClientSession
+
+from metalookup.app.models import Explanation, StarCase
+from metalookup.core.extractor import Extractor, download_tag_lists
+from metalookup.core.website_manager import WebsiteData
+from metalookup.lib.tools import runtime
+
+_FOUND_LIST_MATCHES = "Found list matches"
+_FOUND_NO_LIST_MATCHES = "Found no list matches"
+
+
+class AdBlockBasedExtractor(Extractor[set[str]]):
+    """
+    Base class for features to be extracted with `adblockparser` package.
+
+    The returned extra data corresponds to a sample of the blocked links.
+    The sample size can be configured with the limit member variable.
+    """
+
+    urls: list[str] = []  # needs to be provided by derived class
+
+    def __init__(self, logger: Optional[Logger] = None):
+        self.logger = logger or logging.getLogger(__name__)
+        self.adblock_parser_options = {
+            "script": True,
+            "image": True,
+            "stylesheet": True,
+            "object": True,
+            "xmlhttprequest": True,
+            "object-subrequest": True,
+            "subdocument": True,
+            "document": True,
+            "elemhide": True,
+            "other": True,
+            "background": True,
+            "xbl": True,
+            "ping": True,
+            "dtd": True,
+            "media": True,
+            "third-party": True,
+            "match-case": True,
+            "collapse": True,
+            "donottrack": True,
+            "websocket": True,
+            "domain": "",
+        }
+        self.rules: AdblockRules = None  # noqa will be initialized in async setup call
+        self.limit: Optional[int] = 2
```

Comment:
Why is this optional? How can it be none?

Reply:
While initialized to a value, setting it to None (e.g from outside) is also a valid choice and will (should) be handled correctly in the extract method.
It's not about what type it is in that scenario, it's about what it is allowed to be. Also, the docstring tells you why it is allowed to be None :P